tests/net/tcp/src/main.c

```diff
@@ -2856,4 +2861,145 @@ ZTEST(net_tcp, test_client_seq_validation)
 	k_sleep(K_MSEC(CONFIG_NET_TCP_TIME_WAIT_DELAY));
 }
 
+struct ack_check_struct {
+	int s_data_len;
+	int c_ack;
+	int c_exp_seq;
+};
+
+/* The Congestion Window is by default enabled to be small (e.g. 67B),
+ * so DONT set the s_data_len bigger than it or the expected_seq will
+ * fail the check.
+ */
+static struct ack_check_struct ack_check_list[] = {
+	{10, 20, 10},
+	{20, 9, 30},
+	{30, -2000, 60},
+	{50, 111, 110},
+};
+
+static int ack_check_index;
+static uint32_t expected_seq;
+static uint32_t svr_seq_base;
+static int ackerr;
+
+static void handle_server_ack_validation_test(struct net_pkt *pkt)
+{
+	struct tcphdr th;
+	struct net_pkt *reply;
+	int ret;
+
+	ret = read_tcp_header(pkt, &th);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	if (FL(&th.th_flags, &, PSH)) {
+		/* If server data packet, then return with c_ack in the ack_check_list[] */
+		uint32_t old_ack = ack;
+
+		ack = (uint32_t)(svr_seq_base + ack_check_list[ack_check_index].c_ack);
```

Comment:
We could avoid the outer cast if the `struct ack_check_struct.c_ack` field would be `uint32_t` and then there would be no problem adding signed value to unsigned one